src/v/serde/envelope.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <cinttypes>
+#include <type_traits>
+
+namespace serde {
+
+using version_t = uint8_t;
+
+template<version_t V>
+struct version {
+    static constexpr auto const v = V;
+};
+
+template<version_t V>
+struct compat_version {
+    static constexpr auto const v = V;
+};
+
+/**
+ * \brief provides versioning (version + compat version)
+ *        for serializable aggregate types.
+ *
+ * \tparam Version         the current type version
+ *                         (change for every incompatible update)
+ * \tparam CompatVersion   the minimum required version able to parse the type
+ */
+template<
+  typename T,
+  typename Version,
+  typename CompatVersion = compat_version<Version::v>>
+struct envelope {
+    using envelope_size_t = uint32_t;
+    using value_t = T;
+    static constexpr auto __version = Version::v;
+    static constexpr auto __compat_version = CompatVersion::v;
+};
+
+namespace detail {
+
+template<typename T>
+struct inherits_from_envelope {
+    using Type = std::decay_t<T>;
+    static constexpr auto const value = std::is_base_of_v<
+      envelope<
+        Type,
+        version<Type::__version>,
+        compat_version<Type::__compat_version>>,
+      Type>;
+};
+
+template<typename T, typename = void>
+struct has_compat_attribute : std::false_type {};
+
+template<typename T>
+struct has_compat_attribute<
+  T,
+  std::void_t<decltype(std::declval<T>().__compat_version)>>
+  : std::true_type {};
+
+} // namespace detail
+
+template<typename T>
+inline constexpr auto const is_envelope_v = std::conjunction_v<
+  detail::has_compat_attribute<T>,
+  detail::inherits_from_envelope<T>>;
```

Comment:
1. need this for version, in addition to compat version?
2. going to check if this is still needed
3. perhaps we can statically assert that T does not have members that conflict with our version/compat_version name. this is not a clear thought.

if we choose to keep this I think we need to write some comments here about what's going on.

Reply:
```cpp
std::conjunction_v<
  detail::has_compat_attribute<T>,
  detail::has_version_attribute<T>>;
```

It now just requires `redpanda_serde_compat_version` + `redpanda_serde_compat_version`. So theoretically, you do not need to inherit from `envelope`. Just add those two attributes to your struct, and you're fine. However, using `envelope` could be more future-proof because we can make changes centrally.